oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java

```diff
@@ -0,0 +1,154 @@
+   * @param testData - test data for evaluation
+   * @return Davies Bouldin index - measure of clustering quality, lower the better
+   */
+  static double daviesBouldinIndex(final List<ClusterInfo> clusters, JavaRDD<Vector> testData) {
+    Map<Integer, Tuple2<Double, Long>> clusterSumDistanceAndCounts =
+        fetchClusterSumDistanceAndCounts(clusters, testData).collectAsMap();
+
+    return calcDBIndex(clusterSumDistanceAndCounts, clusters);
+  }
+
+  /**
+   * Computes the Dunn Index of a given clustering, See http://en.wikipedia.org/wiki/Dunn_index
+   * @param clusters - List of {@link ClusterInfo}
+   * @param testData - test data for evaluation
+   * @return Dunn Index - higher the better
+   */
+  static double dunnIndex(final List<ClusterInfo> clusters, JavaRDD<Vector> testData) {
+    List<Tuple2<Integer, Tuple2<Double, Long>>> clusterSumDistanceAndCounts =
+        fetchClusterSumDistanceAndCounts(clusters, testData).collect();
+
+    return calcDunnIndex(clusterSumDistanceAndCounts, clusters);
+  }
+
+  private static double calcDBIndex(Map<Integer, Tuple2<Double, Long>> clusterSumDistAndCounts,
+                                    List<ClusterInfo> clusters) {
+    double totalDBIndex = 0.0;
+    for (int i = 0; i < clusters.size(); i++) {
+      double maxDBIndex = 0;
+      for (int j = 0; j < clusters.size(); j++) {
+        if (i != j) {
+          ClusterInfo c1 = clusters.get(i);
+          ClusterInfo c2 = clusters.get(j);
+
+          if (clusterSumDistAndCounts.containsKey(i) && clusterSumDistAndCounts.containsKey(j)) {
+            double clusterScatter1 = clusterSumDistAndCounts.get(i)._1() / clusterSumDistAndCounts.get(i)._2();
+            double clusterScatter2 = clusterSumDistAndCounts.get(j)._1() / clusterSumDistAndCounts.get(j)._2();
+
+            double dbIndex = (clusterScatter1 + clusterScatter2) / getDistance(c1.getCenter(), c2.getCenter());
+
+            if (dbIndex > maxDBIndex) {
+              maxDBIndex = dbIndex;
+            }
+          }
+        }
+      }
+      totalDBIndex += maxDBIndex;
+    }
+
+    return totalDBIndex / clusters.size();
+  }
+
+  private static double calcDunnIndex(List<Tuple2<Integer,Tuple2<Double,Long>>> clusterSumDistAndCounts,
+                                      List<ClusterInfo> clusters) {
+    double maxIntraClusterDistance = 0.0;
+    for (Tuple2<Integer,Tuple2<Double,Long>> entry : clusterSumDistAndCounts) {
+      double intraClusterDistance = entry._2()._1() / entry._2()._2();
+      if (maxIntraClusterDistance < intraClusterDistance) {
+        maxIntraClusterDistance = intraClusterDistance;
+      }
+    }
+
+    double minInterClusterDistance = Double.POSITIVE_INFINITY;
+    for (int i = 0; i < clusters.size(); i++) {
+      // Distances are symmetric, hence d(i,j) == d(j,i)
+      for (ClusterInfo cluster : clusters) {
+        double distance = getDistance(clusters.get(i).getCenter(), cluster.getCenter());
```

Comment:
Pretty minor here but getting the center could happen outside the inner loop. Is `i` needed in the outer loop or can it also be foreach?